src/Transports.AspNetCore/WebSockets/BaseSubscriptionServer.cs

```diff
@@ -0,0 +1,488 @@
+                    {
+                        disposer = null;
+                    }
+                }
+                finally
+                {
+                    disposer?.Dispose();
+                }
+            }
+            else if (result.Executed)
+            {
+                await SendSingleResultAsync(message, result);
+            }
+            else
+            {
+                await SendErrorResultAsync(message, result);
+            }
+        }
+        catch (OperationCanceledException) when (CancellationToken.IsCancellationRequested)
+        {
+            throw;
+        }
+        catch (ExecutionError error)
+        {
+            if (!Subscriptions.Contains(messageId, dummyDisposer))
+                return;
+            await SendErrorResultAsync(message, error);
+        }
+        catch (Exception ex)
+        {
+            if (!Subscriptions.Contains(messageId, dummyDisposer))
+                return;
+            var error = await HandleErrorDuringSubscribeAsync(message, ex);
+            await SendErrorResultAsync(message, error);
+        }
+    }
+
+    /// <summary>
+    /// Creates an <see cref="ExecutionError"/> for an unknown <see cref="Exception"/>.
+    /// </summary>
+    protected virtual Task<ExecutionError> HandleErrorDuringSubscribeAsync(OperationMessage message, Exception ex)
+        => Task.FromResult<ExecutionError>(new UnhandledError("Unable to set up subscription for the requested field.", ex));
+
+    /// <summary>
+    /// Sends a single result to the client for a subscription or request, along with a notice
+    /// that it was the last result in the event stream.
+    /// </summary>
+    protected virtual async Task SendSingleResultAsync(OperationMessage message, ExecutionResult result)
+    {
+        await SendDataAsync(message.Id!, result);
+        await SendCompletedAsync(message.Id!);
+    }
+
+    /// <summary>
+    /// Sends an execution error to the client during set-up of a subscription.
+    /// </summary>
+    protected virtual Task SendErrorResultAsync(OperationMessage message, ExecutionError executionError)
+        => SendErrorResultAsync(message.Id!, executionError);
+
+    /// <summary>
+    /// Sends an execution error to the client during set-up of a subscription.
+    /// </summary>
+    protected virtual Task SendErrorResultAsync(string id, ExecutionError executionError)
+        => SendErrorResultAsync(id, new ExecutionResult { Errors = new ExecutionErrors { executionError } });
+
+    /// <summary>
+    /// Sends an error result to the client during set-up of a subscription.
+    /// </summary>
+    protected virtual Task SendErrorResultAsync(OperationMessage message, ExecutionResult result)
+        => SendErrorResultAsync(message.Id!, result);
+
+    /// <summary>
+    /// Sends an error result to the client during set-up of a subscription.
+    /// </summary>
+    protected abstract Task SendErrorResultAsync(string id, ExecutionResult result);
```

Comment:
4 overloads for `SendErrorResultAsync` look excessive.

Reply:
Yeah, although they are just overloads and it shouldn't hurt.  There's a reason also: one pair has `OperationMessage` representing the original message that triggered the error response, whereas the other pair just has the id.  In this way you can add logging or whatever from the original message.  And for the second parameter, there's a single error vs a whole ExecutionResult.  Seems to make sense, even if there are a few extra overloads here.

FYI my code uses 3 of the 4 overloads, and the 4th is called by one of the other 3 overloads.  I wouldn't change it.

Reply:
> In this way you can add logging or whatever from the original message.

Yes, but one can override both overloads adding logging and then call `base.SendErrorResultAsync` so logging will be doubled.

Reply:
I mean caller does not know the implementation details. It sees 4 overload for SendErrorResultAsync. What should he do to add logging? I would override all that overloads.

Reply:
Well, actually you would need to override multiple of them anyway, assuming you want to log information that is not present in the abstract method (meaning information from OperationMessage).  That's part of the benefit of having the overloads in the first place.  Let's add comments here, if you like, and move on.

Reply:
You reason from the position that the client knows the implementation. I explain that these additional overloads are confused and I would not say that they have more benefits than harm in this regard. Yes, at least add a note that user should override only last overload with id and result.